wordpress-mobile/WordPress-Android
# Remove ktlint & update detekt to `1.19.0`

Fixes #14012 

**Why?** Ktlint is no friend of the `@Composable` annotation on parameters (see paqN3M-v8-p2), and we plan to start adopting [Jetpack Compose](https://developer.android.com/jetpack/compose) for building UIs.

#### 📃 Changes:
- Remove ktlint task from CI
- Remove ktlint config & dependency
- Update `coding-style` doc to suggest using detekt instead of ktlint
- Replace ktlint comments with detekt suppressions
- Fix `EmptyFunctionBlock` detekt errors by using expression body ( `= UNIT`)
- Suppress `MultiLineIfElse` in `SubfilterPageViewModelTest`
- Suppress `MayBeConst` detekt errors
- Suppress `MaxLineLength` detekt error in `PostUtilsUnitTest`
- Fix and suppress detekt errors in `MySiteViewModelTest` & restore test lost during a merge 
- Bump detekt from `1.15.0` to `1.19.0`
- Migrate detekt reports configuration to the recommended task-level approach 
- Update detekt baseline to suppress 118 existing errors
- Apply suggestions 2, 6 & 7 from the list of 7 in [comment](https://github.com/wordpress-mobile/WordPress-Android/issues/14011#issuecomment-775885924) for #14011

#### 🤔 Questions:

1. @ParaskP7 Is this point still relevant considering we're adding the ktlint ruleset with the `detektPlugins` declaration?
   > When the decommission process for KtLint starts, the responsible person should traverse all the existing KtLint rules and make sure that those are being transferred to Detekt. Step no.1 about would help with that but still there might be cases where some rules might still required to be ported from KtLint to 
Detekt.

   My understanding from two pages in the detekt docs ([formatting rules] (https://detekt.dev/docs/rules/formatting/) & [adding more rule-sets](https://detekt.dev/docs/intro#adding-more-rule-sets)) is that the rules are already included when the `detektPlugins` dependency is added 🤷‍♂️.

2. Why the changes of rules names in `detekt.yml`?
Apparently these have changed with the new version, it was easy enough to figure them out as there were build errors pointing to the deprecated rule names. Mostly this concerns the renames into `ignoreAnnotated`.

3. Is it ok to add yet another 118 suppressions to the `baseline` file?
I started with the idea of fixing all the errors, but they proved to be too many (e.g. most of those are one error per file, which makes the job much more time consuming.
If agreed we can continue with new suppressions in the baseline and consider fixing these rather sooner than later.
I did some exploration and I tend to agree with most of the errors outlined, ideally we'd fix them 😃 .

### 🧪 To test:
- Make sure to read the [detekt changelog](https://detekt.dev/docs/introduction/changelog/#1190---2021-11-29) for the versions greater than `1.15.0` up to `1.19.0` and **check the migration notes against the changes** to verify the migration was done correctly.
- Pull branch `issue/14012-remove-ktlint` locally
- Run `./gradlew WordPress:detekt` and 
  **expect** a successful build.
- Delete `config/detekt/baseline.xml` file
  - run `./gradlew WordPress:detektBaseline`
  - run `git status` and
    **expect** no changes in the `detekt/baseline.xml` file.
- Locate the Detekt-related pipeline job in Buildkite and **expect** success.

#### ⚠️ Merge Instructions:

1. Make sure @wordpress-mobile/owl-team approved the PR (@ParaskP7).
2. Remove the `🔴 Not Ready for Merge` label.
5. Merge 🚀.


## Regression Notes
1. Potential unintended areas of impact
   `N/a`

2. What I did to test those areas of impact (or what existing automated tests I relied on)
   `N/a`

3. What automated tests I added (or what prevented me from doing so)
   `N/a`

PR submission checklist:
